### PR TITLE
Remove scanner from compiler stack

### PR DESCRIPTION
**Pull request by chriseth**
It seems that this can even be further simplified so that the parser constructs the scanner itself.

### CODE REVIEW

**Comment by chriseth** [comment]
Someone can gladly take this over. I don't think it is complicated. The idea is to replace `Scanner` by `CharStream` everywhere except in the parser. The parser can construct an instance of `Scanner` itself which does not have to be long-lived.

File: libsolidity/parsing/Parser.cpp [hunk 1]
```diff
@@ -81,13 +81,14 @@ class Parser::ASTNodeFactory
...
-ASTPointer<SourceUnit> Parser::parse(shared_ptr<Scanner> const& _scanner)
+ASTPointer<SourceUnit> Parser::parse(CharStream& _charStream)
 {
 	solAssert(!m_insideModifier, "");
 	try
 	{
 		m_recursionDepth = 0;
-		m_scanner = _scanner;
+		m_source = &_charStream;
+		m_scanner = make_shared<Scanner>(_charStream);
```
Comment (chriseth) [inline libsolidity/parsing/Parser.cpp:90]: Does this really need a shared pointer?

Comment (chriseth) [reply libsolidity/parsing/Parser.cpp:90]: Oh it kind of does. A `unique_ptr` would probably be better, but for that you need the full Scanner in the header. Although scanner.h is not that big anymore, is it? Maybe `std::optional<Scanner>` could also work.

Comment (christianparpart) [reply libsolidity/parsing/Parser.cpp:90]: Oh I am all in favor of preferring `unique_ptr`, it's just that the whole repo is full of `shared_ptr`'s. I"ll adapt. :-)

Comment (christianparpart) [reply libsolidity/parsing/Parser.cpp:90]: tried and reverted as it's only working if `Scanner` is fully declared for use with `unique_ptr` ;-(

File: libyul/AsmParser.h [hunk 2]
```diff
@@ -89,6 +89,7 @@ class Parser: public langutil::ParserBase
 	/// @param _reuseScanner if true, do check for end of input after the `}`.
 	/// @returns an empty shared pointer on error.
 	std::unique_ptr<Block> parse(std::shared_ptr<langutil::Scanner> const& _scanner, bool _reuseScanner);
+	// TODO: pass CharStream here instead ^^^^^^
```
Comment (chriseth) [inline libyul/AsmParser.h]: todo

**Review by chriseth — COMMENTED** [review]
Approved.

**Pull request merged**